fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewPushDownRewriter.java

```diff
@@ -0,0 +1,857 @@
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rewrite.ReplaceColumnRefRewriter;
+import com.starrocks.sql.optimizer.rule.Rule;
+import com.starrocks.sql.optimizer.rule.transformation.materialization.equivalent.RewriteEquivalent;
+import com.starrocks.sql.optimizer.rule.tree.pdagg.AggColumnRefRemapping;
+import com.starrocks.sql.optimizer.rule.tree.pdagg.AggRewriteInfo;
+import com.starrocks.sql.optimizer.rule.tree.pdagg.AggregatePushDownContext;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.optimizer.OptimizerTraceUtil.logMVRewrite;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.AggregateFunctionRollupUtils.REWRITE_ROLLUP_FUNCTION_MAP;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.AggregateFunctionRollupUtils.genRollupProject;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.AggregateFunctionRollupUtils.getRollupFunctionName;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.MvUtils.deriveLogicalProperty;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.MvUtils.getQuerySplitPredicate;
+
+/**
+ * A mv rewriter that supports to push down aggregate functions below join operator and rewrite the query by mv transparently.
+ */
+public class AggregatedMaterializedViewPushDownRewriter extends MaterializedViewRewriter {
+    private final MvRewriteContext mvRewriteContext;
+    private final ColumnRefFactory queryColumnRefFactory;
+    private final OptimizerContext optimizerContext;
+    private final Rule rule;
+    private final PreVisitor preVisitor = new PreVisitor();
+    private final PostVisitor postVisitor = new PostVisitor();
+
+    public AggregatedMaterializedViewPushDownRewriter(MvRewriteContext mvRewriteContext,
+                                                      OptimizerContext optimizerContext,
+                                                      Rule rule) {
+        super(mvRewriteContext);
+
+        this.mvRewriteContext = mvRewriteContext;
+        this.queryColumnRefFactory = mvRewriteContext.getMaterializationContext().getQueryRefFactory();
+        this.optimizerContext = optimizerContext;
+        this.rule = rule;
+    }
+
+    @Override
+    public OptExpression doRewrite(MvRewriteContext mvContext) {
+        OptExpression input = mvContext.getQueryExpression();
+
+        // try push down
+        AggregatedMaterializedViewPushDownRewriter rewriter =
+                new AggregatedMaterializedViewPushDownRewriter(mvContext, optimizerContext, rule);
+        OptExpression inputDuplicator = duplicateOptExpression(mvContext, input);
```

Comment:
`inputDuplicator` is copied because `AggregatedMaterializedViewPushDownRewriter` is not safe if the rewrite failed, so can return `inputDuplicator` directly if agg push down and rewrite failed.

Reply:
yes, I known the reason why to add Duplicator here, just curious about what if we do it without mofifying column id then we can avoid to add a new project operator.

Reply:
We may do this later?